gps/vcs_source.go

```diff
@@ -203,6 +203,34 @@ func (s *gitSource) exportRevisionTo(ctx context.Context, rev Revision, to strin
 		}
 	}
 
+	// Now do the same for any submodule, using a shell command that "git submodule"
+	// iterates over.
+	{
+		cmd := commandContext(ctx, "git", "submodule", "foreach", "--recursive",
+			// The command gets called in the working tree
+			// of the git submodule.  The .git is a text
+			// file which contains a relative link to the
+			// git directory. We need to follow that and
+			// there do the same "git read-tree + git
+			// checkout-index" with a temporary index as
+			// for the main repo.
+			//
+			// $toplevel is our initial root directory
+			// plus any intermediate git repo (i.e. if
+			// root contains "foo" and "foo" contains
+			// "bar", then we first get called with
+			// "<root>" when exporting "foo" and then with
+			// "<root>/foo" for bar. By stripping our root
+			// directory we get a relative path that we can
+			// append to the target directory.
+			"(cd $(grep gitdir .git | sed -e 's;gitdir: *;;' .git) && mv index origindex && trap 'mv origindex index' EXIT && git read-tree \"$sha1\" && git checkout-index -a --prefix=\""+to+"$(echo \"$toplevel\" | sed -e 's;"+r.LocalPath()+";;')/$path/\")",
```

Comment:
`grep`, `sed`, etc., - not gonna be portable to windows 😢

Let's take an approach with two execs - call out to git to enumerate the submodules, loop over the results with a call to `fs.RenameWithFallback()`, then call out again with `git submodule foreach`. Simplicity, correctness and portability are worth the performance hit.

Reply:
i think you should be able to get what you need from `git submodule status` output. Icky, but it's there.

Reply:
FWIW, you mentioned before (https://github.com/golang/dep/pull/1909#issuecomment-402057799) that the shell code might be okay because "tests pass on appveyor suggests it's probably fine". That was before I added an explicit test for submodules, but it continued to pass also after I added the new testcase. So just to be sure, automatic testing in appveyor includes running dep on Windows?

But I've rewritten the code such that it is based on the output of `git submodule status` anyway, with the "export tree" functionality done by the same Go code as before, just moved to a new method. It's easier to understand this way. The only thing I don't like about it is parsing the output of `git submodule status`, because there is no well-defined, machine-readable output format.